keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,165 @@
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 and `memory.max` in cgroups v2 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs.
+
+Above implement has following disadvantages:
+- `Guaranteed` memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` memory allocation overcommit ( request < limit ) could increase OOM kill possibility because of node memory running out
+
+Currently we only use `memory.limit_in_bytes == limits.memory` to limit memory hardly in cgroups v1 and `memory.max == limits.memory` in cgroups v2.
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
+| memory.min | memory.min specifies a minimum amount of memory the cgroup must always retain, i.e., memory that can never be reclaimed by the system. If the cgroup’s memory usage reaches this low limit and can’t be increased, the system OOM killer will be invoked. |
+
+This proposal uses `memory.min` to retain request memory for `Guaranteed` pods. And for `Burstable` pods, it uses `memory.low` to achieve best-effort memory protection as well as using `memory.high` to throttle allocation of memory overcommit when pods specify `requests.memory`. We also set `memory.max = limits.memory` to do hard limit to make OOM killer work.
+
+### User Stories (Optional)
+- Workload is memory-sensitive, this feature can retain memory to reduce allocation latency
+- Workload is overcommitted in memory, this feature can throttle memory overcommit to increase stability by reducing risk of kernel OOM
+
+### Risks and Mitigations
+- `Guaranteed` pods memory reservation could reduce node memory supply
+- `Burstable` pods memory allocation would slow down when memory use goes over request
+
+## Design Details
```

Comment:
will we also need to update the CRI api? https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L569-L586

CRI api only currently has `memory_limit_in_bytes`. The discussion here is centered around the cgroup config for the pod sandbox, but will it also make sense for each individual container to have it's own memory cgroup v2 settings applied by the underlying container runtime as well?

The discussion here is around s

Reply:
IMO Kubernetes QoS is designed to qos level, e.g. `kubepods/burstable` or `kubepods/besteffort`. We can sink to pod level to set memory qos options, e.g. `kubepods/pod<uid>`. And I think container is not intended to be considered in qos class by kubernetes. What is your opinion?

Reply:
Yup, I think the QoS settings make sense. My comment was more around the fact that as part of this change, we're effectively making use of new cgroup related settings (e.g. `memory.min`). Using those settings on the pod cgroup makes sense, but does it also make sense to propagate those cgroup settings on the container level as well so that for example each guaranteed container has for example `memory.min` == container.memory.request?

Reply:
Agree. Container level limit also makes sense. I think we should create anther KEP to describe this feature.

Reply:
Yes, we will have to expose the knobs at the container level :+1:

Reply:
Maybe we can follow the rules below:
1. if every containers in pod is set `requests.memory` and `limits.memory` ( guaranteed pod ), we set memory.min = requests.memory at container cgroup level
2. if some containers in pod are set `requests.memory` or `limit.memory` ( burstable pod ), we set memory.low = sum(pod.spec.containers.resources.requests[memory]) at pod cgroup level

Reply:
@xiaoxubeii Yeah, that sounds good.

Reply:
Updated